### DISCUSSION TITLE
[Feature Request] Add BTHome support to default tasmota32-bluetooth driver

### DISCUSSION THREAD

**Discussion by AleXSR700** [arendst/Tasmota#20763] [Ideas]
Hi everyone,
@Staars recently added BTHome support to the esp32_mi BLE driver and I was wondering if this could be added to the "normal" bluetooth driver loaded when building using tasmota32-bluetooth config?

Problem is that my ESP32 is being used as a BLE to wifi hub for TRVs and I lose the ability to use any other BLE devices when switching to the BTHome supporting driver.

Would be really cool if it could be added to the normal BLE driver. No problem to custom build of course as long as the driver continues to work with all other BLE devices :)

Thank you for considering.

Alex

EDIT for reference:
https://github.com/arendst/Tasmota/pull/20625

**Comment by AleXSR700** [comment]
Maybe as an additional pro:
pvvx support will be deprecated so the current implementation of mijia support in tasmota32-bluetooth will stop working.

**Comment by AleXSR700** [comment]
Anybody here who could and would help with this?

**Reply by barbudor** [reply]
Can you please explain ?

**Reply by JochenFriedrich** [reply]
See #20625

**Reply by AleXSR700** [reply]
> Can you please explain ?

The BTHome format is the only one that will remain in the future for pvvx's custom firmwares for Xiaomi Mijia devices ( https://github.com/arendst/Tasmota/pull/20625#issuecomment-1972323345 ).

This means that the currently, in Tasmota supported formats will stop working.

Staars already added BTHome support in the linked PR. However, the format has not yet been added to the "normal" Bluetooth driver.
As the smaller one supports hardly anything else (like TRVs etc.), BTHome support is needed on the normal Bluetooth driver.

I tried adding it myself but failed miserably. So I was hoping someone with more skills could solve the problem :-)

**Reply by barbudor** [reply]
Understood
April and May have been crazy-busy at work but I'll try to start looking at it next week end

**Reply by AleXSR700** [reply]
Just fyi because I had linked the initial PR earlier, I just read the release notes for the latest tasmota stable release and it seems Staars updated his initial BTHome implementation in https://github.com/arendst/Tasmota/pull/20989

Please, let me know if you need a tester for BTHome devices. I have a bunch and am happy to flash an esp32 with a test version in case you don't have any BTHome devices (or would simply like a beta tester:) ).

**Reply by AleXSR700** [reply]
Just wanted to check in and see if there is any way I can be of assistance this week? :)

**Reply by AleXSR700** [reply]
Hello barbudor,
did you by any chance have time to take a look at the BTHome implementation? :)

**Reply by AleXSR700** [reply]
Just my keep-alive hoping BTHome support will be added to the tasmota32-bluetooth firmware 🙏

**Reply by AleXSR700** [reply]
👋

**Comment by AleXSR700** [comment]
Push :)

**Comment by DJVirus1980** [comment]
What about add BTHome v2 support?

**Reply by g6094199** [reply]
will be ignored like the reqeusts before....

**Comment by Jason2866** [comment]
Not ignored. Open Source Software is not where you can make wishes and this wishes will be statisfied.
So maybe someone is someday coding this feature and adds it. The idea is open and pushing it every month will not change anything on this.

**Reply by AleXSR700** [reply]
I was just pushing because barbudor commented that he was swamped but would look into it when he had time. Since I cannot expect him to have it on his todo, it is my job to follow up :)

**Reply by Hypfer** [reply]
@AleXSR700 No, it is not "your job" to "follow up".

You were told clearly and unmistakably by a member of the project that you are to cease the spam because that's not how it works.
That barbudor user is not a member of the project. If you feel like you are obliged to ping them, go send them an email. The address is on their profile.

**Reply by AleXSR700** [reply]
I was not disagreeing nor saying that I would continue to do so. I was simply explaining why ;)

**Reply by barbudor** [reply]
Sorry @AleXSR700 I too bury down in work these time, I don't have much time for tasmota and also the BLE stuff is quite compilcated. Not even sure I would succeed.
Let asks to @btsimonh to have a look into.

**Reply by DJVirus1980** [reply]
Hello, has anything changed with the addition of the BTHome protocol?

**Comment by AleXSR700** [comment]
I just saw that all other formats will be dropped with version 6 of ATC Mi thermometer firmware. Is there anybody who wants to try ro implement BTHome with me? If not, all I can do is feed Staars' code into Claude and CoPilot and hope it will be able to implement it for me 🙈

**Reply by g6094199** [reply]
do u use homeassistant? if yes, forgett tasmota and use esphome as BT proxy. switched to it and it works out of the box. even its anoying to waste HA space and time to make own builds for a proxy.

**Reply by AleXSR700** [reply]
esphome has issues when you have a lot of bluetooth devices. It is apparently only stable for a handful of devices. I have 15+.
But yes, was/am considering it. But I then also need to route to MQTT because I use my tasmota to relay the data to an MQTT server that is invisible to Home Assistant, then filter the data and relay it to the MQTT server visible to HA. So I use the proxy as a kind of data router for my filtering.

**Reply by g6094199** [reply]
ok, good to know. i have less than 10 right now.

then maybe openmqttgateway will do the job?

**Reply by AleXSR700** [reply]
> then maybe openmqttgateway will do the job?

I considered that but I am not sure that it actually supports BTHome v2 for our Mi sensors. Maybe they added it now but I think it also did not support in the past

**Reply by g6094199** [reply]
in the forums the setup gets discussed several times. so it should be supported.

**Reply by pkkrusty** [reply]
Esphome can forward bthome data directly to a MQTT server. I had Claude write a yaml for me that does the job. Will post it in the morning.

**Reply by AleXSR700** [reply]
> in the forums the setup gets discussed several times. so it should be supported.

Yes, it seems the discussions I found are outdated. According to the list of compatible devices, BTHome v2 should work as long as the devices have the correct naming scheme (ATC*).

I will need to check it out on one of my esp32 dev kits. Looks doable :)

**Reply by AleXSR700** [reply]
> Esphome can forward bthome data directly to a MQTT server. I had Claude write a yaml for me that does the job. Will post it in the morning.

As mentioned, tried ESPHome before and it was very unstable when running a lot of BLE sensors. It only allows 3 active connections at a time (which should be fine) but used to crash frequently when you had more than 10 BLE devies.

Might also be fixed by now, but will probably try OpenMQTTGateway first (and maybe ESPHome in parallel once that is setup) :)

**Reply by AleXSR700** [reply]
Installed OpenMQTTGateway which was easy enough. Does not have all the flexibility of tasmota but it was easy to setup.

However, when I switch the Mijia sensor to BTHome v2 the OMG can no longer forward the sensor values.

So it seems that BTHome v2 is not working with Mijia sensors in OMG.

Once I switch back to pvvx format, OMG forwards the sensor data again.

Waiting for my account to be approved so I can check in their forum to confirm it was not just a misconfiguration somewhere.

**Reply by g6094199** [reply]
maybe you should open a ticket at the OMG guys for the problem....link it here so i can appy to it

**Reply by AleXSR700** [reply]
I would check in the forum first to make sure I did not miss anything. And then open a ticket.

But I also now tried ESPHome and it seems that it also does not support BTHome v2 as format. It works with pvvx but not BTHome v2.

So result seems to be that BTHome is the new, universal format but it is **not** supported by
- tasmota
- esphome
- openmqttgateway 

So none of the popular choices seem to have BTHome implemented for atc/pvvx firmware yet 🙈

**Reply by g6094199** [reply]
esphome supports bthomev2 for sure.
since my pvvx/THB2 units only support bthomev2:


```
Main features
With default settings:

BLE advertisement interval in [BTHome v2](https://bthome.io/) format is 5 seconds.
```

**Reply by pkkrusty** [reply]
See my post on home assistant forums. Works for me also. I'm sending straight to mqtt server and also home assistant

https://community.home-assistant.io/t/praise-for-ai-assistance-for-yaml/932771

**Reply by AleXSR700** [reply]
Okay, then clearly my config must be off or I completely misunderstood what I need to configure.

My understanding was that I have to add each Mijia sensor as described here: https://esphome.io/components/sensor/xiaomi_ble/#lywsd03mmc

Meaning, since I am using pvvx firmware I would use this:
```yaml
sensor:
  - platform: pvvx_mithermometer
    mac_address: XX:XX:XX:XX:XX:XX
    temperature:
      name: "PVVX Temperature"
    humidity:
      name: "PVVX Humidity"
    battery_level:
      name: "PVVX Battery-Level"
    battery_voltage:
      name: "PVVX Battery-Voltage"
    signal_strength:
      name: "PVVX Signal"
   ```
   
When I do, I only see pvvx formatted broadcasts and not BTHome formatted braodcasts automatically being routed to MQTT with proper, automatic identification of the different sensor data like temp, hum, batt etc..
   
@g6094199 :   I also have THB2 sensors which I am currently not using because not supported by tasmota32-bluetooth driver (due to BTHome v2). But I could not find them in the xiaomi_ble docs of ESPHome. So how do you configure them in ESPHome?

@pkkrusty : I saw your config but does that mean you are manually converting the BTHome data?
```
...
                if (!encrypted && version == 2) {  // BTHome v2, unencrypted
                  // Parse sensor data starting from byte 1 and build JSON
                  std::string json_data = "{";
                  json_data += "\"device\":\"" + device_name + "\",";
                  json_data += "\"mac\":\"" + mac_address + "\",";
                  json_data += "\"rssi\":" + to_string(x.get_rssi()) + ",";
                  json_data += "\"timestamp\":" + to_string(millis()) + ",";

...
                    switch (object_id) {
                      case 0x01: // Battery %
                        if (i < service_data.data.size()) {
                          sensor_key = "battery";
                          sensor_value = to_string(service_data.data[i++]);
                          sensor_unit = "%";
                        }
                        break;
 ```


Edit:
Claude.ai and CoPilot both say that
`You're correct - that is a limitation. With the bthome platform in ESPHome, you do need to manually specify each entity (temperature, humidity, battery, etc.) for each device. `
If that is the case, then ESPHome does not really support BTHome format for Xiaomi mijia devices. Because for pvvx format you only specify the device (based on MAC) and the platform automatically identifies the different entities, just like tasmota does.
If ESPHome only allows manual creation of each individual entity, then okay, I could do that, but then it would be far inferior to having tasmota support BTHome format (because there you do not even have to define the devices, it just automatically identifies them based on their names).

**Reply by pkkrusty** [reply]
Yes the esphome config is manually formatting for mqtt, but simply acting as a Bluetooth proxy means home assistant picks up the bthome broadcasts automatically. My config does the manual parsing because I need it in a specific format. FWIW, ai generated the config for me, so it may not be optimal. But it works. @AleXSR700

**Reply by g6094199** [reply]
as u can see this config is way old, but since its working flawlessly there was no need to adapt.....

its was simple as that:

> esphome:
> 
>   name: esp32-bluetooth-proxy
>   friendly_name: Bluetooth Proxy
>   min_version: 2024.6.0
>   name_add_mac_suffix: true
> 
> esp32:
>   board: esp32dev
>   framework:
>     type: esp-idf
> 
> 
> logger:
> 
> 
> esp32_ble_tracker:
>   scan_parameters:
>     # We currently use the defaults to ensure Bluetooth
>     # can co-exist with WiFi In the future we may be able to
>     # enable the built-in coexistence logic in ESP-IDF
>     active: true
> 
> bluetooth_proxy:
>   active: true
> 
> button:
>   - platform: safe_mode
>     id: button_safe_mode
>     name: Safe Mode Boot
> 
>   - platform: factory_reset
>     id: factory_reset_btn
>     name: Factory reset
> 
> # Enable Home Assistant API
> api:
>   encryption:
>     key: "dWw="
> 
> ota:
>   - platform: esphome
>     password: "b8"
> 
> wifi:
>   ssid: !secret wifi_ssid
>   password: !secret wifi_password
> 
>   # Enable fallback hotspot (captive portal) in case wifi connection fails
>   ap:
>     ssid: "Ble-Gw Fallback Hotspot"
>     password: "B"
>

**Reply by AleXSR700** [reply]
@g6094199:
Wait, so you are not sending the ble data to mqtt? You are just bridging between bluetooth device and HA?
If so, then ESPHome is **not** doing anything with the data and the format becomes irrelevant as long as HA knows it.
ESPHome is essentially nothing more than a repeater.

If that is the case, then that is a completely different use case. In my use case, the esp32 "decodes" the incoming ble data and converts it to a define syntax. The esp32 already breaks it down into devices and entities automatically. Essentially, the esp32 does what the BTHome integration in HA does for you.

@pkkrusty :
Seems like a similar problem in your case. Rather than ESPHome actually doing the decoding itself, you are doing that manually and then forwarding the reformatted information. Again, not what tasmota does.

Maybe this screenshot of the tasmota WebUI frontend clears things up:
![Screenshot_20251026_221106_Chrome.jpg](https://github.com/user-attachments/assets/e7095cc0-6256-4213-a9b7-2e454e4cec06)

Tasmota identifies Xiaomi Mijia TH devices automatically (without me telling it anything) and identifies the device and its entities and knows what they represent. It decodes the incoming information and makes it user readable.

That is what "support BTHome v2 format on pvvx firmware devices" means. :)

**Reply by pkkrusty** [reply]
I'm with you. I looked to tasmota to do the job at first also, since I prefer the ecosystem. But in the absence of bthomev2 support, I am using ESPHome to manually create a compatible json structure so I can ingest data into influxdb using telegraf.

**Reply by AleXSR700** [reply]
Yeah, giving up on alternatives because none currently support BTHome with Mijia devices.

Tried using Claude.ai and CoPilot to implement it into the xsns 62 driver but no luck so far.

**Reply by pkkrusty** [reply]
Maybe worth trying again with Opus 4.6...it's pretty smart.

**Comment by Jason2866** [comment]
Had a look at openmqttgateway. Since the BLE part is maintained from h2zero (the author of the NimBLE libs) it is probably not a bad choice ;-) The list of supported BLE devices is long.

**Reply by AleXSR700** [reply]
Planned now as next test. Means I cannot use my power socket with esp32 anymore but that would be fine. :)

https://github.com/arendst/Tasmota/discussions/20763#discussioncomment-14777007

**Reply by AleXSR700** [reply]
List of supported BLE devices by tasmota is longer :)

But I am stuck anyway, because it seems none of the big three support BTHome for atc/pvvx firmware.

https://github.com/arendst/Tasmota/discussions/20763#discussioncomment-14783965

**Reply by DJVirus1980** [reply]
Can you add this devises https://github.com/pvvx/THB2

**Reply by DJVirus1980** [reply]
<img width="1069" height="1598" alt="изображение" src="https://github.com/user-attachments/assets/6b09d231-2f23-40ab-8019-59286ca96592" />
It is work.

**Reply by DJVirus1980** [reply]
Whith custom xsns_62_esp32_mi_ble.ino

**Reply by g6094199** [reply]
> custom xsns_62_esp32_mi_ble.ino

hi,
is there a github link?

**Reply by DJVirus1980** [reply]
Now testing.

**Reply by AleXSR700** [reply]
> Now testing.

Any update? Did you figure it out?

**Reply by DJVirus1980** [reply]
no, its works. link [xsns_62_esp32_mi_ble.zip](https://github.com/user-attachments/files/26190304/xsns_62_esp32_mi_ble.zip)

**Comment by DJVirus1980** [comment]
[xsns_62_esp32_mi_ble.zip](https://github.com/user-attachments/files/26190304/xsns_62_esp32_mi_ble.zip)
Testing in dev 15.3.0.3

**Reply by DJVirus1980** [reply]
It works 3 months. may be add to project?

**Reply by pkkrusty** [reply]
Works for me in initial testing. However I note in the driver that you only implemented a few of the sensor types. Do you plan on adding the rest or do you want me to continue? 

When I say me, I mean Claude.

**Reply by AleXSR700** [reply]
Would be nice to add to official repo. If I, and by I I also mean Claude or GPT, can help, let me know. But it seems to work, so maybe open a PR?

**Reply by pkkrusty** [reply]
I should clarify, BTHome v2 protocol defines about 25 different sensor types, including binary sensors. In my case the BTH-01 temp/humidity sensor can upload a different firmware from pvvx which has support for binary sensor on an exposed pad, which in my case I put on a reed switch so it can act as a door sensor as well. But it doesn't show up on the tasmota GUI even though it is transmitting the hex header. Home Assistant picks it up correctly via bluetooth.

**Reply by arendst** [reply]
Pls try latest dev branch which includes BTHome support for some sensors and possibly your binary sensor too.

**Comment by arendst** [comment]
I've merged the changes from the .zip file in latest dev `xsns_62_esp32_mi_ble.ino`.

I hoped to be able to decode my IoTorero WS01 BTHome V2 button but all I see is:
```
15:12:48.030 M32: BTHome: encrypted packets not yet supported (41E47BC3077EAAB05C922A3A4D2141273C160000806CBE96)
15:12:48.055 M32: BTHome: encrypted packets not yet supported (41E47BC3077EAAB05C922A3A4D2141273C160000806CBE96)
15:12:48.111 M32: BTHome: encrypted packets not yet supported (41E47BC3077EAAB05C922A3A4D2141273C160000806CBE96)
```

Pity.

**Comment by arendst** [comment]
And now with decryption:
```
18:08:48.131 M32: BTHome: MAC `0C3D5E54C3E9` encrypted packet (41EE37ABF296FA3E5574285CAC6A3D47C81800006DC52D24)
18:08:48.134 M32: E9C3545E3D0C: Searching key
18:08:48.145 M32: E9C3545E3D0C: Decryption key found
18:08:48.146 M32: E9C3545E3D0C: Decrypted 635F97030063C3580100638BAEAC00
18:08:48.148 M32: BTHome: decrypted packet (41635F97030063C3580100638BAEAC00)
18:08:48.161 M32: 0C3D5E54C3E9: Payload decrypted
18:08:48.162 M32: 0C3D5E54C3E9: Frame 0, last 0
18:08:48.163 M32: BTHome: unknown obj 0x63 - stopping
18:08:48.174 M32: BTHome: 0C3D5E54C3E9 slot 0
```

TBC...

**Comment by arendst** [comment]
BTHome button decoding works too:
```
12:50:15.249 MQT: stat/atomlite2/SENSOR = {"Time":"2026-08-04T12:50:15","BTHome54c3e9":{"mac":"0c3d5e54c3e9","Btn":1,"Battery":72,"RSSI":-57}}
```

**Comment by DJVirus1980** [comment]
<img width="1146" height="1763" alt="изображение" src="https://github.com/user-attachments/assets/528219d3-ca54-40b4-b0f3-2d48eef64e06" />
Hi, It works.
